src/core/server/saved_objects/export/saved_objects_exporter.ts

```diff
@@ -0,0 +1,154 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import Boom from '@hapi/boom';
+import { createListStream } from '@kbn/utils';
+import { PublicMethodsOf } from '@kbn/utility-types';
+import { SavedObject, SavedObjectsClientContract } from '../types';
+import { fetchNestedDependencies } from './fetch_nested_dependencies';
+import { sortObjects } from './sort_objects';
+import {
+  SavedObjectsExportResultDetails,
+  SavedObjectExportBaseOptions,
+  SavedObjectsExportByObjectOptions,
+  SavedObjectsExportByTypeOptions,
+} from './types';
+
+/**
+ * @public
+ */
+export type ISavedObjectsExporter = PublicMethodsOf<SavedObjectsExporter>;
+
+/**
+ * @public
+ */
+export class SavedObjectsExporter {
+  private savedObjectsClient: SavedObjectsClientContract;
+  private exportSizeLimit: number;
+
+  constructor({
+    savedObjectsClient,
+    exportSizeLimit,
+  }: {
+    savedObjectsClient: SavedObjectsClientContract;
+    exportSizeLimit: number;
+  }) {
+    this.savedObjectsClient = savedObjectsClient;
+    this.exportSizeLimit = exportSizeLimit;
+  }
+
+  /**
+   * Generates an export stream for given types.
+   *
+   * See the {@link SavedObjectsExportByTypeOptions | options} for more detailed information.
+   */
+  public async exportByTypes(options: SavedObjectsExportByTypeOptions) {
+    const objects = await this.fetchByTypes(options);
+    return this.processObjects(objects, options);
+  }
+
+  /**
+   * Generates an export stream for given object references.
+   *
+   * See the {@link SavedObjectsExportByObjectOptions | options} for more detailed information.
+   */
+  public async exportByObjects(options: SavedObjectsExportByObjectOptions) {
+    if (options.objects.length > this.exportSizeLimit) {
+      throw Boom.badRequest(`Can't export more than ${this.exportSizeLimit} objects`);
+    }
+    const objects = await this.fetchByObjects(options);
+    return this.processObjects(objects, options);
+  }
+
+  private async processObjects(
+    rootObjects: SavedObject[],
+    {
+      excludeExportDetails = false,
+      includeReferencesDeep = false,
+      namespace,
+    }: SavedObjectExportBaseOptions
+  ) {
+    let exportedObjects: Array<SavedObject<unknown>> = [];
+    let missingReferences: SavedObjectsExportResultDetails['missingReferences'] = [];
+
+    if (includeReferencesDeep) {
+      const fetchResult = await fetchNestedDependencies(
+        rootObjects,
+        this.savedObjectsClient,
+        namespace
+      );
+      exportedObjects = sortObjects(fetchResult.objects);
+      missingReferences = fetchResult.missingRefs;
+    } else {
+      exportedObjects = sortObjects(rootObjects);
+    }
+
+    // redact attributes that should not be exported
+    const redactedObjects = exportedObjects.map<SavedObject<unknown>>(
+      ({ namespaces, ...object }) => object
```

Comment:
question: can we collocate this logic with filtering `score` out to avoid spreading the logic all over the file?

Reply:
`fetchByObjects` uses `bulkGet`, which returns raw `SavedObjects` without score.
`fetchByTypes` uses `find`, which returns `SavedObjectsFindResult` that are just SO + `score` field

I think moving the score handling logic outside of the 'by type' specific logic would complexify more than simplify the code, as we would need to handle `SavedObjects | SavedObjectsFindResult` in `processObjects`?

Reply:
We can create a normalization function per an export method to remove all the unnecessary fields at one pass. Feel free to ignore the suggestion if it doesn't justify the complexity.